packages/pagination/src/index.ts

```diff
@@ -215,3 +218,73 @@ export const createPagination = (
 
   return [paginationProps, page, setPage as Setter<number>];
 };
+
+declare module "solid-js" {
+  namespace JSX {
+    interface Directives {
+      loaderDirective: boolean
+    }
+  }
+}
+
+export type _E = JSX.Element;
+
+type loaderDirective = (ref: Element) => void
+
+/**
+ * ```typescript
+ * const [pages, asLoader, { page, setPage, setPages, end, setEnd }] = createInfiniteScroll(fetcher);
+ * ```
+ * @param fetcher (page: number) => Promise<T[]>
+ * @return `pages()` is an accessor contains array of contents 
+ * @property `pages.loading` is a boolean indicator for the loading state
+ * @property `pages.error` contains any error encountered
+ * @return `asLoader` is a directive used to set loader
+ * @method `page` is an accessor that contains page number
+ * @method `setPage` allows to manually change the page number
+ * @method `setPages` allows to manually change the contents of the page
+ * @method `end` is a boolean indicator for end of the page
+ * @method `setEnd` allows to manually change the end
+ */
+export function createInfiniteScroll<T>(
+  fetcher: (page: number) => Promise<T[]>,
+): [
+    pages: Accessor<T[]>,
+    asLoader: loaderDirective,
+    options: {
+      page: Accessor<number>,
+      setPage: Setter<number>,
+      setPages: Setter<T[]>
+      end: Accessor<boolean>,
+      setEnd: Setter<boolean>,
+    }
+  ] {
+  const [loader, setLoader] = createSignal<HTMLElement>();
+  const [pages, setPages] = createSignal<T[]>([]);
+  const [page, setPage] = createSignal(0);
+  const [end, setEnd] = createSignal(false);
+
+  const visible = createVisibilityObserver()(loader);
+  const [contents] = createResource(page, fetcher);
+
+  createEffect(() => {
+    if (visible() && !end() && !contents.loading) {
+      setPage(p => p + 1);
+    }
+  })
+
+  createEffect(() => {
+    const content = contents();
+    if (!content) return;
+    if (content.length === 0) setEnd(true);
+    setPages(p => [...p, ...content!]);
+  })
```

Comment:
Could this be done with derivations instead of synchronizing the state with effects?
If the effect is the only way, then `createComputed` will be better.